src/source/stdin.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2020-2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#![cfg(not(tarpaulin_include))]
+
+use crate::{source::prelude::*, url::TremorUrl};
+use async_std::io;
+
+pub struct Stdin {
+    onramp_id: TremorUrl,
+}
+
+pub struct Int {
+    onramp_id: TremorUrl,
+    origin_uri: EventOriginUri,
+    stdin: io::Stdin,
+}
+
+impl std::fmt::Debug for Int {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "stdin")
+    }
+}
+
+impl Int {
+    async fn new(uid: u64, onramp_id: TremorUrl) -> Result<Self> {
+        let stdin = io::stdin();
+        let origin_uri = EventOriginUri {
+            uid,
+            scheme: "tremor-stdin".to_string(),
+            host: hostname(),
+            port: None,
+            path: vec![],
+        };
+        Ok(Self {
+            onramp_id,
+            origin_uri,
+            stdin,
+        })
+    }
+}
+
+impl onramp::Impl for Stdin {
+    fn from_config(id: &TremorUrl, _config: &Option<YamlValue>) -> Result<Box<dyn Onramp>> {
+        Ok(Box::new(Self {
+            onramp_id: id.clone(),
+        }))
+    }
+}
+
+#[async_trait::async_trait]
+impl Onramp for Stdin {
+    async fn start(&mut self, config: OnrampConfig<'_>) -> Result<onramp::Addr> {
+        let source = Int::new(config.onramp_uid, self.onramp_id.clone()).await?;
+        SourceManager::start(source, config).await
+    }
+
+    fn default_codec(&self) -> &str {
+        "string"
+    }
+}
+
+#[async_trait::async_trait()]
+impl Source for Int {
+    fn id(&self) -> &TremorUrl {
+        &self.onramp_id
+    }
+
+    async fn pull_event(&mut self, _id: u64) -> Result<SourceReply> {
+        let mut input = String::new();
+        if let Ok(len) = self.stdin.read_line(&mut input).await {
```

Comment:
We shouldn't split the input into lines here, that's the job of the `lines` preprocessor. That way it mirrors the behaviour of the `stdout` sink where the `lines` postprocessor adds the newlines to the output (if required)

Reply:
I don't think `read_line()` splits the input into lines. For example: passing an input like `snot \n badger`, would be treated as one single string "snot \n badger", which is similar to how the websocket onramp works for example. Do let me know if I'm missing something here :D

Reply:
The [documentation](https://docs.rs/async-std/0.99.7/async_std/io/struct.Stdin.html#method.read_line) says "Reads a line of input" so I'm fairly sure it does read a line.

The second issue, enforcing UTF8 compliance as it uses a `String` also is relevant.

Reply:
Done! ✅ @Licenser

Reply:
nice :D thank you!